src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -65,20 +67,32 @@ def __init__(
   def _combined_invalidating_fileset_from_globs(glob_strs, root):
     return set.union(*(Fileset.globs(glob_str, root=root)() for glob_str in glob_strs))
 
+  def _get_snapshot(self, glob):
+    """Returns a Snapshot of the input glob"""
+    return self._scheduler.capture_snapshots(
```

Comment:
Rather than using `capture_snapshots`, which supports capturing a snapshot from an arbitrary location (but does _not_ support caching the result), you'll want to use:
```
return scheduler.product_request(Snapshot, subjects=[PasthGlobs(globs)])[0]
```
...which will cache most of the execution run over run.

You'll just need to move the call to `self._maybe_invalidate_scheduler_batch` in `self._handle_batch_event` _after_ the call to `self._scheduler.invalidate_files`, so that we notice immediately whether the digest has changed.